### PR TITLE
Add Binary instances to Field and Enumeration types

### PR DESCRIPTION
**Pull request by yigitozkavci**
Deriving `Binary` instance for `Field` and `Enumeration` types are almost trivial and creating instances as a pinch library user results in having orphan instances. This PR derives those instances.

An unrelated note: It looks like `pinch.cabal` is manually modified in one of the previous commits and results in a warning when running `stack repl`. I didn't fix it in this PR as it results in large unrelated diff. Let me know if you'd like it fixed though, I can open a separate PR for it.

### CODE REVIEW

**Comment by abhinav** [comment]
Hey, thanks for the PR! This looks fine but would you please clarify the use
case? I'm concerned that Binary instances here would mislead users into
believing that they can get a Thrift-compatible binary representation through
it.

**Comment by yigitozkavci** [comment]
> Would you please clarify the use case?

Sure! I've been working on https://github.com/yigitozkavci/parquet-hs for reading Parquet files which uses Thrift definitions for representing their metadata. This library itself indeed doesn't need any Binary instances.

I'm also working on integrating parquet-hs to http://github.com/utdemir/distributed-dataset, which will send Parquet metadata definitions to distribute reading parts of Parquet data in arbitrary backends. One possible backend is Amazon lambda, for instance.

Long story short, I'll be sending these Thrift definitions through arbitrary network channels, which is why I needed the Binary instances.

If you think this will do more harm than good for the future users of this repo, I can either silence the orphan warnings or use a fork on my end. It probably isn't worth adding if it will confuse the users.

**Comment by abhinav** [comment]
I'm not strongly opposed to merging this but as I said, I suspect users will misinterpret the Binary instance on their types. As you mentioned, silencing the warnings could be a valid option here. Alternatively, since these are Thrift objects, there's also the option of encoding them into Thrift payloads using the compact or binary protocol and sending that over the wire.

**Comment by phile314** [comment]
I am taking the liberty of closing this as it looks unlikely that any progress is going to happen here. In any case, using orphan instances looks like an appropriate solution. Feel free to re-open.
